aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -268,6 +290,300 @@ std::pair<at::Tensor, std::vector<int64_t*>> pack_vecs(
+  ) {
+    uint4 val = zero;
+    if(is_aligned<uint4>(src + i)) {
+      stream_load128(val, src + i);
+    } else {
+      for (size_t j = 0; j < BYTES_PER_THREAD; ++j) {
+        reinterpret_cast<char*>(&val)[j] = src[i + j];
+      }
+    }
+    stream_store128(&dst[i], val);
+  }
+  if(thread_idx < align_off && thread_idx < max_chunk_size) {
+    char val = (char) 0;
+    if (thread_idx < actual_chunk_size) {
+      val = src[thread_idx];
+    }
+    dst[thread_idx] = val;
+  }
+  while(align_end + thread_idx < max_chunk_size) {
+    char val = (char) 0;
+    if (align_end + thread_idx < actual_chunk_size) {
+      val = src[align_end + thread_idx];
+    }
+    dst[align_end + thread_idx] = val;
+    align_end += num_threads;
+  }
+}
+
+// Read `actual_chunk_size` bytes from bfloat16 array `src`  (i.e., actual_chunk_size/2 bfloat16 elements),
+// cast bfloat16 to float, and write to float array `dst`. Pad 0.0f if 2*actual_chunk_size < max_chunk_size.
+// src: pointer to a bfloat16 array
+// dst: pointer to a float32 array
+// max_chunk_size: number of bytes in dst array
+// actual_chunk_size: number of bytes in src array, which is less than or equal to max_chunk_size / 2
+static __device__ __inline__ void copy_chunk_with_pad_cast_bfloat16_to_float(
+  char* dst,
+  const char* src,
+  int64_t max_chunk_size,
+  int64_t actual_chunk_size,
+  int64_t thread_idx,
+  int64_t num_threads
+) {
+  const int64_t dst_elem_size = 4;
+  const int64_t src_elem_size = 2;
+  const int64_t max_num_elems = max_chunk_size / dst_elem_size;
+  const int64_t actual_num_elems = actual_chunk_size / src_elem_size;
+  int64_t elem_index = thread_idx;
+  while (elem_index < actual_num_elems) {
+#if defined(USE_ROCM) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800))
+    reinterpret_cast<float*>(dst)[elem_index] = static_cast<at::BFloat16>(
+      reinterpret_cast<const at::BFloat16*>(src)[elem_index]
+    );
+#else
+    reinterpret_cast<float*>(dst)[elem_index] = __bfloat162float(
+      reinterpret_cast<const __nv_bfloat16*>(src)[elem_index]
+    );
+#endif
+    elem_index += num_threads;
+  }
+  while (elem_index < max_num_elems) {
+    reinterpret_cast<float*>(dst)[elem_index] = 0.0f;
+    elem_index += num_threads;
+  }
+}
+
+// NOTE [CUDA kernel for chunk_cat]
+// chunk_cat_cuda adopts a "jagged grid" strategy, inspired by NOTE [CUDA fast path for split_with_sizes_copy.out].
+// In addition, chunk_cat_cuda supports padding via copy_chunk_with_pad when src chunk size is less than
+// dst chunk size.
+static __global__ void chunk_cat_cuda_kernel(
+  char** src,
+  char* dst,
+  int64_t* block_idx_to_tensor_idx,
+  int64_t* tensor_idx_to_start_tensor_bytes,
+  int64_t* start_block_idx_per_tensor_chunk,
+  int64_t* actual_tensor_sizes,
+  int64_t* pad_tensor_chunk_sizes,
+  int64_t* num_blocks_per_tensor_chunk,
+  int64_t slice_size,
+  int64_t chunk_size,
+  bool bf162float // copying from bf16 to float if True else assuming same elem type for src and dst
```

Comment:
I'm not sure why you need this? If you only ever support some dtypes, do you just want to template over the dtypes you support here?

Reply:
+1. I feel if we already have a variant that supports casting, there shouldn't be a no-cast variant since it's just a special case of casting to the same type.